File: DIANN/DIANN_quant_P2.py
#!/usr/bin/env python3
"""
Quick DIA-NN MaxLFQ Extraction
Simple script to extract MaxLFQ values from filtered DIA-NN output
"""

import pandas as pd
import numpy as np
import sys
from pathlib import Path

def extract_maxlfq(input_file):
    """Extract MaxLFQ values from DIA-NN parquet file"""
    
    print("Quick DIA-NN MaxLFQ Extraction")
    print("-" * 50)
    
    # Load data
    print(f"Loading: {input_file}")
    df = pd.read_parquet(input_file)
    
    print(f"Loaded {len(df):,} rows")
    print(f"Proteins: {df['Protein.Group'].nunique():,}")
    print(f"Samples: {df['Run'].nunique()}")
    
    # Get sample names
    sample_mapping = {}
    for run in df['Run'].unique():
        sample_name = Path(run).stem
        # Clean up common suffixes
        for suffix in ['_DIA', '_dia', '.raw', '.mzML', '.d']:
            sample_name = sample_name.replace(suffix, '')
        sample_mapping[run] = sample_name
    
    print("\nSample names:")
    for orig, clean in list(sample_mapping.items())[:5]:
        print(f"  {clean}")
    if len(sample_mapping) > 5:
        print(f"  ... and {len(sample_mapping) - 5} more")
    
    # Extract MaxLFQ values
    print("\nExtracting PG.MaxLFQ values...")
    
    # Check if PG.MaxLFQ exists
    if 'PG.MaxLFQ' not in df.columns:
        print("ERROR: PG.MaxLFQ column not found!")
        print("Available columns:", df.columns.tolist())
        return
    
    # Pivot to get protein x sample matrix
    result = df.pivot_table(
        index='Protein.Group',
        columns='Run', 
        values='PG.MaxLFQ',
        aggfunc='first'  # Take first value if duplicates exist
    )
    
    # Rename columns to clean sample names
    result.columns = [sample_mapping[col] for col in result.columns]
    
    # Replace zeros with NaN
    result = result.replace(0, np.nan)
    
    # Get protein annotations
    print("Adding protein annotations...")
    annotations = df[['Protein.Group', 'Protein.Names', 'Genes']].drop_duplicates('Protein.Group')
    
    # Merge annotations with quantification
    final = annotations.merge(result, left_on='Protein.Group', right_index=True, how='right')
    final = final.set_index('Protein.Group')
    
    # Move annotation columns to the front
    info_cols = ['Protein.Names', 'Genes']
    sample_cols = [col for col in final.columns if col not in info_cols]
    final = final[info_cols + sample_cols]
    
    # Calculate statistics
    numeric_cols = final.select_dtypes(include=[np.number]).columns
    final['N_Samples'] = final[numeric_cols].notna().sum(axis=1)
    final['Mean_Intensity'] = final[numeric_cols].mean(axis=1)
    final['CV_%'] = (final[numeric_cols].std(axis=1) / final[numeric_cols].mean(axis=1) * 100)
    
    # Save output
    output_file = input_file.replace('.parquet', '_MaxLFQ.csv')
    print(f"\nSaving to: {output_file}")
    final.to_csv(output_file)
    
    # Summary statistics
    print("\n" + "=" * 50)
    print("SUMMARY")
    print("=" * 50)
    print(f"Quantified proteins: {len(final):,}")
    print(f"Samples: {len(sample_cols)}")
    print(f"Missing values: {final[numeric_cols].isna().sum().sum():,} ({final[numeric_cols].isna().sum().sum() / (len(final) * len(numeric_cols)) * 100:.1f}%)")
    print(f"Median CV: {final['CV_%'].median():.1f}%")
    print(f"Proteins in all samples: {(final['N_Samples'] == len(sample_cols)).sum():,}")
    
    print("\nDone!")
    
def main():
    if len(sys.argv) < 2:
        print("Usage: python quick_maxlfq.py <filtered_parquet_file>")
        print("\nExample:")
        print("  python quick_maxlfq.py DR12_merged_experiment_report_filtered.parquet")
        sys.exit(1)
    
    input_file = sys.argv[1]
    
    if not input_file.endswith('.parquet'):
        print(f"Warning: Input file '{input_file}' doesn't have .parquet extension")
    
    try:
        extract_maxlfq(input_file)
    except Exception as e:
        print(f"\nError: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    main()

File: DIANN/DIANN_quant_P2.R
#!/usr/bin/env Rscript
# Quick DIA-NN MaxLFQ Extraction
# Simple script to extract MaxLFQ values from filtered DIA-NN output

# 安装缺失的包
install.packages("tibble")
library(tibble)

# 然后重新运行
maxlfq_data <- extract_maxlfq("filtered_data.parquet")

if (!require(pacman)) install.packages("pacman")
pacman::p_load(arrow, dplyr, tidyr, readr, stringr)

extract_maxlfq <- function(input_file) {
  cat("Quick DIA-NN MaxLFQ Extraction\n")
  cat(strrep("-", 50), "\n")
  
  # Load data
  cat("Loading:", input_file, "\n")
  df <- read_parquet(input_file)
  
  cat("Loaded", formatC(nrow(df), format = "d", big.mark = ","), "rows\n")
  cat("Proteins:", formatC(length(unique(df$Protein.Group)), format = "d", big.mark = ","), "\n")
  cat("Samples:", length(unique(df$Run)), "\n")
  
  # Create sample mapping
  sample_mapping <- df %>%
    distinct(Run) %>%
    mutate(Sample_Name = str_remove_all(basename(Run), "\\.(raw|mzML|d)$") %>%
             str_remove_all("_(DIA|dia)$")) %>%
    deframe()
  
  cat("\nExtracting PG.MaxLFQ values...\n")
  
  if (!"PG.MaxLFQ" %in% names(df)) {
    stop("ERROR: PG.MaxLFQ column not found!")
  }
  
  # Pivot to matrix
  result <- df %>%
    select(Protein.Group, Run, PG.MaxLFQ) %>%
    filter(!is.na(PG.MaxLFQ), PG.MaxLFQ > 0) %>%
    pivot_wider(names_from = Run, values_from = PG.MaxLFQ, values_fn = first) %>%
    column_to_rownames("Protein.Group")
  
  # Rename columns
  names(result) <- sample_mapping[names(result)]
  
  # Add annotations
  annotations <- df %>%
    select(Protein.Group, Protein.Names, Genes) %>%
    distinct()
  
  final <- annotations %>%
    left_join(result %>% rownames_to_column("Protein.Group"), by = "Protein.Group") %>%
    column_to_rownames("Protein.Group")
  
  # Calculate statistics
  numeric_cols <- final %>% select_if(is.numeric) %>% names()
  final$N_Samples <- rowSums(!is.na(final[numeric_cols]))
  final$Mean_Intensity <- rowMeans(final[numeric_cols], na.rm = TRUE)
  final$CV_percent <- apply(final[numeric_cols], 1, function(x) {
    if (sum(!is.na(x)) < 2) return(NA)
    sd(x, na.rm = TRUE) / mean(x, na.rm = TRUE) * 100
  })
  
  # Save output
  output_file <- str_replace(input_file, "\\.parquet$", "_MaxLFQ.csv")
  write_csv(final, output_file)
  
  cat("\nSaved to:", output_file, "\n")
  cat("Done!\n")
  return(final)
}

main <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  if (length(args) < 1) {
    cat("Usage: Rscript DIANN_quant_P2.R <filtered_parquet_file>\n")
    quit(status = 1)
  }
  extract_maxlfq(args[1])
}

if (!interactive()) {
  main()
}